src/EnergyPlus/MixedAir.cc

```diff
@@ -637,56 +649,32 @@ void SimOAComponent(EnergyPlusData &state,
                                                                                       CoolingActive,
-                                                                                      0,
+                                                                                      zoneOAUnitNum,
                                                                                       OAUCoilOutTemp,
                                                                                       ZoneEquipFlag,
                                                                                       sensOut,
                                                                                       latOut);
         }
         OACoolingCoil = true;
     } else if (CompTypeNum == SimAirServingZones::CompType::CoilSystemWater) { // "CoilSystem:Cooling:Water"
-        if (state.dataAirLoop->OutsideAirSys(OASysNum).compPointer[CompIndex] == nullptr) {
-            UnitarySystems::UnitarySys thisSys;
-            state.dataAirLoop->OutsideAirSys(OASysNum).compPointer[CompIndex] =
-                thisSys.factory(state, DataHVACGlobals::UnitarySys_AnyCoilType, CompName, false, 0);
-            UnitarySystems::UnitarySys::checkUnitarySysCoilInOASysExists(state, CompName, 0);
-        }
         if (Sim) {
-            bool HeatingActive = false;
-            bool CoolingActive = false;
-            Real64 OAUCoilOutTemp = 0.0;
-            bool ZoneEquipFlag = false;
-            Real64 sensOut = 0.0;
-            Real64 latOut = 0.0;
+            int compNum = state.dataAirLoop->OutsideAirSys(OASysNum).ComponentIndex(CompNum);
             state.dataAirLoop->OutsideAirSys(OASysNum).compPointer[CompIndex]->simulate(state,
                                                                                         CompName,
                                                                                         FirstHVACIteration,
                                                                                         AirLoopNum,
-                                                                                        CompIndex,
+                                                                                        compNum,
                                                                                         HeatingActive,
                                                                                         CoolingActive,
-                                                                                        CompIndex,
+                                                                                        zoneOAUnitNum,
                                                                                         OAUCoilOutTemp,
                                                                                         ZoneEquipFlag,
                                                                                         sensOut,
```

Comment:
Now here we are inside SimOAComponents. CompIndex is used as in index to the compPointer array as well as passed as an index to the object index. This call would simulate the CoilSystemWater object and pass back as a reference the object index into CompIndex at old line 686. This would then overwrite the value stored in state.dataAirLoop->OutsideAirSys(OASysNum).ComponentIndex(CompNum) on the return from this function. Using a local here avoids this issue. Another possible solution would be to set all valid component indexes in GetOutsideAirSysInputs and not pass as reference the CompIndex, but that seemed to be a bigger change.

Reply:
But that is ultimately what we need to do. Set up all objects and cross-indices in one or two passes and then use value indices from then on. The current approach is inefficient, redundant, and (as you've just seen but I didn't even anticipate) booby-trapped.

Reply:
OK, I did start down that path and saw how many valid components are allowed in the OA system. I agree that is the best approach. Maybe it won't be as bad as first thought.

Reply:
I wouldn't do this quite yet. This probably needs to be part of a "pass".

Reply:
At the very least, there are a good number of other things I would do before I try to tackle this.